### PR TITLE
Add links to patterns for stop deletion

### PR DESCRIPTION
**Pull request by philip-cline**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JSDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

Link a user to the pattern that uses a specific stop when trying to delete it for faster (and much more usable) deletion work flow

### CODE REVIEW

File: lib/common/actions/index.js [hunk 1]
```diff
@@ -131,14 +132,56 @@ function getErrorMessageFromJson (
     }
     // re-assign error message after it gets used in the detail.
     errorMessage = json.message || JSON.stringify(json)
+    if (json.detail && json.detail.includes('Referenced pattern IDs:')) {
```
Comment (miles-grant-ibigroup) [inline lib/common/actions/index.js]: You're essentially creating your own text-based protocol, which is a massive hack. This is fine for now, but we definitely need to add a comment here (including line numbers) that this needs to go and be replaced with a better solution, namely a new endpoint that returns exactly this information in JSON format

File: lib/common/actions/index.js [hunk 2]
```diff
@@ -131,14 +132,56 @@ function getErrorMessageFromJson (
...
 }
 
+const PatternLinkErrorMessage = (props) => {
+  const url = window.location.href
+  const basePath = url.split('edit/')[0] // Remove the current component from the URL
+  const {detail, patterns} = props
+
+  // Limit the list of patterns to 5 to not overwhelm users
+  const unDisplayedPatterns = patterns.length - 5
+  const limitedPatterns = patterns.slice(0, 5)
```
Comment (miles-grant-ibigroup) [inline lib/common/actions/index.js]: I would not do this

File: lib/common/actions/index.js [hunk 3]
```diff
@@ -131,14 +132,56 @@ function getErrorMessageFromJson (
...
+
+  // $FlowFixMe
+  return <div>
+    {detail}
+    {limitedPatterns.map((pattern, index) => {
+      if (index > 5) return
```
Comment (miles-grant-ibigroup) [inline lib/common/actions/index.js]: We don't need this check twice

File: lib/common/actions/index.js [hunk 4]
```diff
@@ -131,14 +132,56 @@ function getErrorMessageFromJson (
...
+    }
     if (json.detail && json.detail.includes('conflicts with an existing trip id')) {
       action = 'DEFAULT'
     }
   }
-  console.error(errorMessage)
   return { action, detail, message: errorMessage }
 }
 
+const PatternLinkErrorMessage = (props) => {
```
Comment (miles-grant-ibigroup) [inline lib/common/actions/index.js:165]: Again, I would add a comment here saying that this entire block needs to be removed when the better solution is put in place. I would reference this PR as well so we remember what we were thinking

File: lib/common/actions/index.js [hunk 5]
```diff
@@ -131,14 +132,56 @@ function getErrorMessageFromJson (
...
+        const patterns = patternsMatch.split(',')
+        // $FlowFixMe
+        detail = <PatternLinkErrorMessage detail={detail} patterns={patterns} />
+      }
+    }
     if (json.detail && json.detail.includes('conflicts with an existing trip id')) {
       action = 'DEFAULT'
     }
   }
-  console.error(errorMessage)
```
Comment (miles-grant-ibigroup) [inline lib/common/actions/index.js]: Why did you remove this? I found it helpful to have the errors in the console

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
This is a fine solution on a time crunch, but we need something cleaner in place for the long run

File: lib/common/actions/index.js (hunk 1 above)
Comment (philip-cline) [reply lib/common/actions/index.js]: Agreed.

File: lib/common/actions/index.js (hunk 2 above)
Comment (philip-cline) [reply lib/common/actions/index.js]: You wouldn't limit the number of patterns shown ? Or just wouldn't slice?

Comment (miles-grant-ibigroup) [reply lib/common/actions/index.js]: For the proper implementation we could maybe do a scrollable list? For now since this is a duct tape solution I would say show them all

Comment (philip-cline) [reply lib/common/actions/index.js]: We could probably do a scrollable list within this PR

Comment (philip-cline) [reply lib/common/actions/index.js]: Ok I added some scrolling, maybe doesn't look the best but we can expand that too in the future.

Comment (philip-cline) [reply lib/common/actions/index.js]: Indeed. Removed in https://github.com/ibi-group/datatools-ui/pull/943/commits/5a027d784a5ab84f546bc645f448d4771f77e9bc

File: lib/common/actions/index.js (hunk 4 above)
Comment (philip-cline) [reply lib/common/actions/index.js:165]: Comment added. Having a component in here itself kind of seems like a sign of the end times.

File: lib/common/actions/index.js (hunk 5 above)
Comment (philip-cline) [reply lib/common/actions/index.js]: Sorry, accident. Restored in https://github.com/ibi-group/datatools-ui/pull/943/commits/5a027d784a5ab84f546bc645f448d4771f77e9bc

File: lib/common/actions/index.js [hunk 6]
```diff
@@ -148,37 +148,48 @@ function getErrorMessageFromJson (
...
+      borderWidth: '2px',
+      maxHeight: '150px',
+      margin: '10px 0px',
+      overflow: 'auto',
+      padding: '5px 10px'
+    }}>
+      {patterns.map((pattern, index) => {
+        // Pattern contains route and pattern data in format {pattern-route}
+        pattern = pattern.slice(1, -1) // Remove curly braces
+        const [patternId, routeId] = pattern.split('-')
```
Comment (miles-grant-ibigroup) [inline lib/common/actions/index.js]: Just to future-proof this, would it be possible to move these lines out of this component and pass it in as props instead?

File: lib/common/actions/index.js [hunk 7]
```diff
@@ -132,9 +132,9 @@ function getErrorMessageFromJson (
     }
     // re-assign error message after it gets used in the detail.
     errorMessage = json.message || JSON.stringify(json)
+    // This section (L135 - L146) is a hack to display links to patterns referenced by a stop to delete.
```
Comment (miles-grant-ibigroup) [inline lib/common/actions/index.js:136]: Can we add a TODO: REMOVE HACK above this line? Just to make it clear this has to go

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
Almost there! Just a few nitpicks

File: lib/common/actions/index.js (hunk 6 above)
Comment (philip-cline) [reply lib/common/actions/index.js]: ur wish is my command https://github.com/ibi-group/datatools-ui/pull/943/commits/1203494986f3560476ac9251715238cb5c4a9fc1

File: lib/common/actions/index.js (hunk 7 above)
Comment (philip-cline) [reply lib/common/actions/index.js:136]: Nothing I would love more: https://github.com/ibi-group/datatools-ui/pull/943/commits/1203494986f3560476ac9251715238cb5c4a9fc1

File: lib/common/actions/index.js [hunk 8]
```diff
@@ -131,6 +132,23 @@ function getErrorMessageFromJson (
...
+      const patternSplit = json.detail.split(/\[(.*?)\]/) // Match text within square brackets (our comma separated list of patternIds)
+      detail = patternSplit[0]
+      const patternsMatch = patternSplit[1]
+      if (patternsMatch) {
+        const patterns = patternsMatch.split(',').map(pattern => {
+          pattern = pattern.slice(1, -1) // Remove curly braces
+          const [patternId, routeId] = pattern.split('-')
+          return {patternId, routeId}
+        })
+        // $FlowFixMe
```
Comment (miles-grant-ibigroup) [inline lib/common/actions/index.js]: It feels terrifying to use a fake patterns object like this -- maybe we should add a comment? Or fix the flow types

**Review by miles-grant-ibigroup — APPROVED** [review]
Works great! Thanks for all the changes

File: lib/common/actions/index.js (hunk 8 above)
Comment (philip-cline) [reply lib/common/actions/index.js]: umm https://github.com/ibi-group/datatools-ui/pull/943/commits/f9f2ba5e447b30a1dfcdd59e7eee31bdeaa62964

File: lib/common/actions/index.js [hunk 9]
```diff
@@ -131,6 +132,22 @@ function getErrorMessageFromJson (
     }
     // re-assign error message after it gets used in the detail.
     errorMessage = json.message || JSON.stringify(json)
+    // TODO: REMOVE HACK
+    // This section (L135 - L146) is a hack to display links to patterns referenced by a stop to delete.
+    // This should be replaced with a better solution such as a new endpoint for this information.
+    if (json.detail && json.detail.includes('Referenced pattern IDs:')) {
```
Comment (binh-dam-ibigroup) [inline lib/common/actions/index.js]: The searched string is inconsistent with the one defined at https://github.com/conveyal/gtfs-lib/blob/30754ba741ee414fc1778837160d91e12694eddf/src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java#L1616, so this will never show the links.
```suggestion
    if (json.detail && json.detail.includes('Referenced patterns:')) {
```

File: lib/common/actions/index.js [hunk 10]
```diff
@@ -139,6 +156,44 @@ function getErrorMessageFromJson (
...
+      maxHeight: '150px',
+      margin: '10px 0px',
+      overflow: 'auto',
+      padding: '5px 10px'
+    }}>
+      {patterns.map((pattern, index) => {
+        const patternPath = `edit/route/${pattern.routeId}/trippattern/${pattern.patternId}`
+        return (
+          <li style={{listStyle: 'none', margin: '3px 0px'}}>
+            <a href={basePath + patternPath}>
```
Comment (binh-dam-ibigroup) [inline lib/common/actions/index.js]: I'd be curious if the link href could be a relative URL.

**Review by binh-dam-ibigroup — APPROVED** [review]
Approving subject to making the referenced patterns string consistent with what is sent from gtfs-lib.

File: lib/common/actions/index.js (hunk 9 above)
Comment (philip-cline) [reply lib/common/actions/index.js]: Good catch the two PRs were out of sync. Updated in https://github.com/ibi-group/datatools-ui/pull/943/commits/78fc07f63988a31c5f525297b54e2db1f33d2916

File: lib/common/actions/index.js (hunk 10 above)
Comment (philip-cline) [reply lib/common/actions/index.js]: Great idea! I updated it to be a relative URL in https://github.com/ibi-group/datatools-ui/pull/943/commits/78fc07f63988a31c5f525297b54e2db1f33d2916
